### PR TITLE
Dw 2426

### PR DESCRIPTION
**Pull request by UdayChokkam**
This code reads all object keys in a given prefix of S3 bucket
- Generate pairs of data and metadata keys from the given list of keys
- Throws fatal error when metadata for a data file doesnt exist
-Logs error and proceeds when data file doesnt exist for a given metadata file

### CODE REVIEW

File: src/main/kotlin/app/batch/KeyPairGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            KeyPair(dataKey, metadatakey)
+        }
+        validateKeyPairs(keyPairs)
+        return keyPairs.filter { it.dataKey !== null && it.metadataKey != null }
+    }
+
+    fun validateKeyPairs(keyPairs: List<KeyPair>) {
+        keyPairs.forEach {
+            if (it.dataKey != null && it.metadataKey == null) {
+                logger.error("Metadata file not found for the data file ${it.dataKey}")
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/KeyPairGenerator.kt]: Make a variable of the text is the same?

File: src/main/kotlin/app/configuration/SecureHttpClientProvider.kt [hunk 2]
```diff
@@ -13,21 +13,21 @@ import javax.net.ssl.SSLContext
...
-class SecureHttpClientProvider: HttpClientProvider {
+class SecureHttpClientProvider : HttpClientProvider {
 
     override fun client(): CloseableHttpClient =
-        HttpClients.custom().run {
-            setDefaultRequestConfig(requestConfig())
-            setSSLSocketFactory(connectionFactory())
-            build()
-        }
+            HttpClients.custom().run {
```
Comment (mark-r-m) [inline src/main/kotlin/app/configuration/SecureHttpClientProvider.kt]: I'd expecting this to be an indent of 4 not 8?

File: src/test/kotlin/app/batch/KeyPairGeneratorTest.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        val captor = argumentCaptor<ILoggingEvent>()
+        verify(mockAppender, times(8)).doAppend(captor.capture())
+        val formattedMessages = captor.allValues.map { it.formattedMessage }
+        assertTrue(formattedMessages.contains("adb.collection.0000.json.gz.pdf,adb.collection.0000.json.gz.docx matched file format but not matched neither data nor metadata file extension"))
+        assertEquals(expectedKeyPairs, actualKeyPairs)
+
+    }
+
+    @Test
+    fun Should_Log_When_Data_File_Doesnt_But_Metadata_File_Exist() {
```
Comment (mark-r-m) [inline src/test/kotlin/app/batch/KeyPairGeneratorTest.kt]: I suggest "Should_Log_When_Data_File_Doesn_Exist_But_Metadata_File_Does" or "Should_Log_When_Data_File_Doesnt_But_Metadata_File_Does_Exist" ??

File: src/test/kotlin/app/batch/KeyPairGeneratorTest.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+        listOf(KeyPair("adb.collection.0000.json.gz.enc", null),
+                KeyPair("cdb.collection.0000.json.gz.enc", "cdb.collection.0000.json.gz.encryption.json"),
+                KeyPair("sdb.collection.0000.json.gz.enc", "sdb.collection.0000.json.gz.encryption.json"))
+
+        val keys = listOf("adb.collection.0000.json.gz.enc", "cdb.collection.0000.json.gz.enc", "sdb.collection.0000.json.gz.enc",
+                "cdb.collection.0000.json.gz.encryption.json",
+                "sdb.collection.0000.json.gz.encryption.json")
+
+        keyPairGenerator.generateKeyPairs(keys, fileFormatRegex, dataFileExtensionRegex, metadataFileExtensionRegex)
```
Comment (mark-r-m) [inline src/test/kotlin/app/batch/KeyPairGeneratorTest.kt:140]: Can we please assert the exception message, to be sure we are not leaking data, and what will be in our logs is useful?

File: src/test/kotlin/app/batch/S3ReaderTest.kt [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        s3Object1.objectContent = S3ObjectInputStream(ByteArrayInputStream(OBJECT_CONTENT1.toByteArray()), HttpGet())
+
+
+        given(s3Client.listObjectsV2(BUCKET_NAME1, S3_PREFIX_FOLDER)).willReturn(listObjectsV2Result)
+        given(s3Client.getObject(BUCKET_NAME1, VALID_DATA_KEY)).willReturn(s3Object1)
+
+        given(s3Client.listObjectsV2(BUCKET_NAME1, S3_PREFIX_FOLDER)).willReturn(listObjectsV2Result)
+        given(s3Client.getObject(BUCKET_NAME1, VALID_DATA_KEY)).willReturn(s3Object1)
+
+        s3Reader.read()
```
Comment (mark-r-m) [inline src/test/kotlin/app/batch/S3ReaderTest.kt:135]: Can we please assert the exception message, to be sure we are not leaking data, and what will be in our logs is useful?

**Review by mark-r-m — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/batch/KeyPairGenerator.kt (hunk 1 above)
Comment (UdayChokkam) [reply src/main/kotlin/app/batch/KeyPairGenerator.kt]: I  have now moved  it to variable

**Review by mark-r-m — APPROVED** [review]

File: src/main/kotlin/app/configuration/SecureHttpClientProvider.kt (hunk 2 above)
Comment (UdayChokkam) [reply src/main/kotlin/app/configuration/SecureHttpClientProvider.kt]: I have changed  all continuation indentation to be 4 across files

File: src/test/kotlin/app/batch/KeyPairGeneratorTest.kt (hunk 3 above)
Comment (UdayChokkam) [reply src/test/kotlin/app/batch/KeyPairGeneratorTest.kt]: Modified to "Should_Log_When_Data_File_Doesnt_But_Metadata_File_Does_Exist" now

File: src/test/kotlin/app/batch/KeyPairGeneratorTest.kt (hunk 4 above)
Comment (UdayChokkam) [reply src/test/kotlin/app/batch/KeyPairGeneratorTest.kt:140]: Asserted the log message

File: src/test/kotlin/app/batch/S3ReaderTest.kt (hunk 5 above)
Comment (UdayChokkam) [reply src/test/kotlin/app/batch/S3ReaderTest.kt:135]: I have delegated all the pairing logic to keypairgenerator and all logs are asserted in its corresponding test keypairgeneratortest

**Pull request merged**